runtime/shared_heap.c

```diff
@@ -785,6 +804,470 @@ void caml_verify_heap(caml_domain_state *domain) {
+
+        if (Has_status_val(v, NOT_MARKABLE)
+          || Whsize_val(v) > SIZECLASS_MAX ) return;
+
+        /* look up the pool v is in and check if we're evacuating */
+        if( caml_pool_of_shared_block(v)->evacuating == 1 ) {
+          value fwd = Field(v, 0);
+          CAMLassert( caml_pool_of_shared_block(fwd)->evacuating == 0 );
+          CAMLassert(Is_block(fwd));
+          /* We need to point not to the forwarded Closure though but to the
+            matching Infix_tag inside of it */
+          CAMLassert(Tag_val(fwd + infix_offset) == Infix_tag);
+          *p = fwd + infix_offset;
+        }
+      } else {
+        mlsize_t vsize = Whsize_hd(vhd);
+
+        if( vsize <= SIZECLASS_MAX )
+        {
+          if( caml_pool_of_shared_block(v)->evacuating == 1) {
+            value fwd = Field(v,0);
+            CAMLassert( caml_pool_of_shared_block(fwd)->evacuating == 0 );
+            CAMLassert(Is_block(fwd));
+            /* Update p to point to the first field of v */
+            *p = fwd;
+            /* Check the original block and the forwarded block
+                are the same size */
+            CAMLassert(Whsize_val(*p) == vsize);
+          }
+        }
+      }
+    }
+  }
+
+static void compact_update_block(value* p) {
+  header_t hd = Hd_hp(p);
+
+  /* We should never be called with a block that has a zero header (this would
+    indicate a bug in traversing the shared pools). */
+  CAMLassert(hd != 0);
+
+  tag_t tag = Tag_hd(hd);
+
+  mlsize_t wosz = Wosize_hd(hd);
+  mlsize_t i;
+  uintnat offset = 0;
+
+  /* We should never encounter an Infix tag iterating over the shared pools or
+    large allocations. We could only find it in roots by those use
+    [compact_update_field]. */
+  CAMLassert( tag != Infix_tag );
+
+  if( tag == Cont_tag ) {
+    value stk = Field(Val_hp(p), 0);
+    if (Ptr_val(stk) != NULL) {
+      caml_scan_stack(&compact_update_field, 0, NULL, Ptr_val(stk), 0);
+    }
+  } else {
+    if( tag == Closure_tag ) {
+      offset = Start_env_closinfo(Closinfo_val(Val_hp(p)));
+    }
+
+    if( tag < No_scan_tag ) {
+      for( i = offset; i < wosz; i++ ) {
+        compact_update_field(NULL, Field(Val_hp(p), i), &Field(Val_hp(p), i));
+      }
+    }
+  }
+}
+
+static void compact_update_ephe_list(value ephe_list) {
+  while( ephe_list != 0 ) {
+    mlsize_t wosize = Wosize_val(ephe_list);
+
+    if( Ephe_link(ephe_list) != 0 ) {
+      compact_update_field(NULL, Field(ephe_list, CAML_EPHE_LINK_OFFSET),
+        &Field(ephe_list, CAML_EPHE_LINK_OFFSET));
+    }
+
+    compact_update_field(NULL, Field(ephe_list, CAML_EPHE_DATA_OFFSET),
+      &Field(ephe_list, CAML_EPHE_DATA_OFFSET));
+
+    for( int i = CAML_EPHE_FIRST_KEY ; i < wosize ; i++ ) {
+      compact_update_field(NULL, Field(ephe_list, i), &Field(ephe_list, i));
+    }
+
+    ephe_list = Ephe_link(ephe_list);
+  }
+}
+
+/* Heap compaction */
+void caml_compact_heap(caml_domain_state* domain_state, void* data,
```

Comment:
a caml_gc_log here to indicate that a compaction is going on would be good, so that it shows up in the CAMLRUNPARAM-activated GC tracing.

Reply:
Whoops! I've fixed that (and added runtime_events too) in https://github.com/ocaml/ocaml/pull/12193/commits/4c53583472fb2b1c8706ee7b24892f1c21df8862